### DISCUSSION TITLE
How to find elements within a custom web component?

### DISCUSSION THREAD

**Discussion by nygardryan** [EvolvedWeb/evowc#30] [Q&A] [answered]
I've been doing some experimenting, and I was unable to find any documentation on how to find elements within a custom web component. Here's my code, I'm thinking it would be beneficial to have a section on how to do this:

```
<component tag="color-me" :color_1="str:red">
  <template>
    <input :value="color_1">
  </template>

  <style>
    input {
      color: red
    }
  </style>

  <script>
    update(property) {

      let target = this.querySelector('input')
      console.log(target)
      target.style.color = this.color_1

    }
  </script>
</component>
```

The this.querySelector('input') never returns anything. I understand this probably has to do with the shadowdom, just curious what the appropriate way to do this is?

**Comment by intervalia** [comment, marked as answer]
I am working on getting all of the documentation updated. Until then:

There are two options for this.

1. You could write it like this:

```html
<component tag="color-me" :color_1="str:red" :#myStyle>
  <template>
    <input :value="color_1" :style="#myStyle">
  </template>
  <script>
    update(property) {
      this.#myStyle = `color:${this.color_1}`;
    }
  </script>
</component>
```

In this case you are setting the value of the `#myStyle` CPA and that is auto-bound to the style of the `<input>` tag.

2. Evo auto-tracks all elements that have any kind of binding. The elements in the component are all given an attribute `el` with a unique value. The attribute will look something like `<input el="el1" />`. All of these elements are stored in the class's private property `this.#els` and the generated setter code uses that to connect to the bound elements.

If you want to access an element then you can set our own `el` attribute on that element like this `<p el="mine"></p>`. You can access that element by using the property `this.#els.mine`.

> This is only recommended for doing things that Evo can't do, like calling `focus()` or `select()`. It is best to use CPAs and event handlers for almost everything. This will simplify your code and improve performance.

<input el="myInput" :value="#input" /> could be accessed like this: this.#els.myInput.focus();

Remember that Evo does not regenerate the UI. It just changes properties and attributes on the existing elements.
